### DISCUSSION TITLE
The NewCounterFunc doesn't allow using labels on metrics

### DISCUSSION THREAD

**Discussion by ppatierno** [prometheus/client_golang#924] [Q&A]
While working on the Strimzi canary project [1] I had this code at the beginning to track a metric related to produced records:

```go
recordsProduced = promauto.NewCounterVec(prometheus.CounterOpts{
			Name:      "records_produced_total",
			Namespace: "strimzi_canary",
			Help:      "The total number of records produced",
		}, []string{"clientid", "partition"})


// at some point in the code
labels := prometheus.Labels{
			"clientid":  ps.canaryConfig.ClientID,
			"partition": strconv.Itoa(i),
		}
		recordsProduced.With(labels).Inc()
```

Now I have to make this counter to be accessible so I was thinking to use a var and then using `NewCounterFunc` so something like this:

```go
var recordsProducedCounter uint64 = 0

recordsProduced = promauto.NewCounterFunc(prometheus.CounterOpts{
		Name:      "records_produced_total",
		Namespace: "strimzi_canary",
		Help:      "The total number of records produced",
	}, func() float64 {
		return float64(atomic.LoadUint64(&recordsProducedCounter))
	})

// at some point in the code
atomic.AddUint64(&recordsProducedCounter, 1)
```

The problem I have now is that I cannot attach updated labels on the metric. The code updates just the variable of course but maybe the func should help about the labels.
Is there any other way to achieve what I need?

[1] https://github.com/strimzi/strimzi-canary

**Comment by ppatierno** [comment]
I saw some reference here https://github.com/prometheus/client_golang/issues/486 but what about the need for labels?

**Comment by beorn7** [comment]
If you know the labels in advance, you can use const labels:

```Go
	recordsProduced = promauto.NewCounterFunc(prometheus.CounterOpts{
		Name:      "records_produced_total",
		Namespace: "strimzi_canary",
		Help:      "The total number of records produced",
		ConstLabels: prometheus.Labels{
			"clientid":  ps.canaryConfig.ClientID,
			"partition": strconv.Itoa(i),
		},
	}, func() float64 {
		return float64(atomic.LoadUint64(&recordsProducedCounter))
	})
```

See also https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#example-GaugeFunc-ConstLabels

If the situation is more dynamic, you have to implement a custom `prometheus.Collector` that internally uses `NewConstMetric` and friends. This example might be helpful: https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#example-Collector

**Reply by ppatierno** [reply]
@beorn7 first of all thanks for the answer!
My case is dynamic so is there any specific reason why the `CounterFunc` wasn't implemented in order to provide a `With(labels)` as the `CounterVec`.
Developing a collector from scratch sounds reasonable but too much just because of the need for setting labels values dynamically that should be quite "normal" to do.

**Comment by beorn7** [comment]
In my experience, implementing the Collector is not much longer (but arguably easier to understand).

Historically, client_golang had way too many top-level types. So I tried to justify any addition of a new top level type with a significant benefit. The CounterFunc and GaugeFunc are elegant and simple and frequently used. Doing the same for complex metric types like Histogram or for any metric vector (AKA "with labels") would be not as simple, arguably less elegant than the Collector approach, and (presumably) not used so often. It would "pollute" the package namespace quite a bit for a rather niche demand (those few people that need something like a "CounterVecFunc" _and_ are keen on saving a few lines of code compared to the Collector approach).
